server/src/main/java/org/elasticsearch/index/mapper/KeyedJsonAtomicFieldData.java

```diff
@@ -0,0 +1,230 @@
+        private final SortedSetDocValues delegate;
+
+        /**
+         * The first and last ordinals whose term has 'key' as a prefix.
+         */
+        private final long minOrd;
+        private final long maxOrd;
+
+        public static KeyedJsonDocValues create(String key, SortedSetDocValues delegate) {
+            BytesRef keyBytes = new BytesRef(key);
+
+            long minOrd, maxOrd;
+            try {
+                minOrd = findMinOrd(keyBytes, delegate);
+                maxOrd = minOrd != -1 ? findMaxOrd(keyBytes, delegate) : -1;
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+
+            return new KeyedJsonDocValues(keyBytes, delegate, minOrd, maxOrd);
+        }
+
+        private KeyedJsonDocValues(BytesRef key,
+                                   SortedSetDocValues delegate,
+                                   long minOrd,
+                                   long maxOrd) {
+            this.key = key;
+            this.delegate = delegate;
+            this.minOrd = minOrd;
+            this.maxOrd = maxOrd;
+        }
+
+        @Override
+        public long getValueCount() {
+            return delegate.getValueCount();
+        }
+
+        /**
+         * Returns the (un-prefixed) term value for the requested ordinal.
+         *
+         * Note that this method can only be called on ordinals returned from {@link #nextOrd()}.
+         * Otherwise it may attempt to look up values that do not share the correct prefix, which
+         * can result in undefined behavior or an error.
+         */
+        @Override
+        public BytesRef lookupOrd(long ord) throws IOException {
+            BytesRef keyedValue = delegate.lookupOrd(ord);
+            int prefixLength = key.length + 1;
+            int valueLength = keyedValue.length - prefixLength;
+            return new BytesRef(keyedValue.bytes, prefixLength, valueLength);
+        }
+
+        @Override
+        public long nextOrd() throws IOException {
+            if (minOrd < 0) {
+                return NO_MORE_ORDS;
+            }
+
+            for (long ord = delegate.nextOrd(); ord != NO_MORE_ORDS; ord = delegate.nextOrd()) {
+                if (minOrd <= ord) {
+                    if (ord <= maxOrd) {
+                        return ord;
+                    } else {
+                        return NO_MORE_ORDS;
+                    }
+                }
+            }
+            return NO_MORE_ORDS;
+        }
+
+        @Override
+        public boolean advanceExact(int target) throws IOException {
+            if (minOrd < 0) {
+                return false;
+            }
+
+            if (delegate.advanceExact(target)) {
+                for (long ord = delegate.nextOrd(); ord != NO_MORE_ORDS; ord = delegate.nextOrd()) {
+                     if (minOrd <= ord && ord <= maxOrd) {
+                         boolean advanced = delegate.advanceExact(target);
```

Comment:
Calls to advanceExact() are often followed by calls to nextOrd(). With such an implementation, we will iterate ords of the underlying doc values twice. Could we somehow cache the first ord that is greater than or equal to `minOrd` so that we don't have to call `advanceExact` again on the underlying doc values here and the first call to nextOrd() will return the cached first matching ord?

Reply:
This is a nice idea, will try it out.